hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/server/StorageContainerManager.java

```diff
@@ -1158,6 +1158,8 @@ public String getDatanodeRpcPort() {
    */
   @Override
   public void start() throws IOException {
+    upgradeFinalizer.runPrefinalizeStateActions(scmStorageConfig, this);
```

Comment:
Question:
So init will successfully happen but the startup will fail?


Because if init happens every time after the upgrade, ratis sever might have started and already initialized ratis server, and wait for leader election.

Do we want to run these checks in init also to avoid ratis server initialization.

Reply:
In the current implementation init will succeed but start will fail. Init does not need to be run in the new version before the new bits are started, but it shouldn't hurt either. I did not realize we were also starting a ratis server in SCM init, so we should probably run the upgrade actions there too, yes. This shouldn't affect new SCMs since they will be finalized and the actions will not run.

Reply:
Took a closer look and init actually cannot happen in a pre-finalized cluster. SCM must already be inited for the cluster to be pre-finalized, and running init again will cause an error in `Storage#setClusterId`. So we shouldn't have to worry about ratis server being started in init for pre-finalized clusters. In new clusters it is not a problem since there is nowhere to downgrade to.

Reply:
Yes, but if init happens again after we upgrade bits with ha enable flag on, then this will not catch the issue right? As on an upgraded cluster, init can happen, check else block of scmInit.
Like in some deployments where init is run with every restart.

Reply:
>>Took a closer look and init actually cannot happen in a pre-finalized cluster.
What is meant by this?
Let's say freshly installed cluster, can we perform init or not?

Reply:
Thanks for the offline discussion @bharatviswa504. I now see that we can run init again after the upgrade, even though it is not required, and must run the check there as well to prevent Ratis setup steps from happening. The PR has been updated accordingly.

Reply:
Just a question why do we need to delay this check until the start? Why can't it be done in the StorageManager constructor for fail fast?

Reply:
This specific check could probably be moved to the constructor, but in general, the upgrade framework's action execution gives a fully constructed service to each action to inspect for conditions which may violate pre-finalize state. See `UpgradeFinalizer#runPrefinalizeStateActions` and its implementations. With the current upgrade action implementation we had to work around this a bit to run this action statically in init, but circumventing the upgrade framework and manually checking for pre-finalized state in every action is not recommended. In the future we may introduce a new type of pre-finalize validation action that can be run statically if these sort of init checks become common.